### PR TITLE
Implement Asm.make_elf

### PR DESCRIPTION
**Pull request by david942j**
Half part of #66

---
This change is [Reviewable](https://reviewable.io/reviews/peter50216/pwntools-ruby/88)

### CODE REVIEW

File: lib/pwnlib/asm.rb [hunk 1]
```diff
@@ -86,6 +97,60 @@ def asm(code)
...
+    #   The load address for the ELF file.
+    #   If +nil+ is given, default address will be used.
+    #   See {DEFAULT_VMA}.
+    # @param [Boolean] to_file
+    #   Returns ELF content or the path to the ELF file.
+    #   If +true+ is given, the ELF will be saved into a temp file, and the path is returned.
+    #   Otherwise, the content of ELF is returned.
+    #
+    # @return [String]
+    #   If +to_filet+ is +false+ (default), returns the content of ELF. Otherwise, a file is created and the path is
```
Comment (david942j) [inline lib/pwnlib/asm.rb]: typo

File: lib/pwnlib/asm.rb [hunk 2]
```diff
@@ -86,6 +97,60 @@ def asm(code)
...
+    #   implementation, this let us have higher flexibility and less binary dependencies.
+    #
+    # @example
+    #   path = make_elf(asm(shellcraft.cat('/proc/self/maps')), to_file: true)
+    #   puts `#{path}`
+    #   # 08048000-08049000 r-xp 00000000 fd:01 27671233                           /tmp/pwn20180129-3411-7klnng.elf
+    #   # f77c7000-f77c9000 r--p 00000000 00:00 0                                  [vvar]
+    #   # f77c9000-f77cb000 r-xp 00000000 00:00 0                                  [vdso]
+    #   # ffda6000-ffdc8000 rwxp 00000000 00:00 0                                  [stack]
+    def make_elf(data, vma: nil, to_file: false)
```
Comment (david942j) [inline lib/pwnlib/asm.rb]: Provide block that accepts path as argument, and remove the file created after yield.

File: lib/pwnlib/asm.rb [hunk 3]
```diff
@@ -12,6 +14,15 @@ module Pwnlib
   module Asm
     module_function
 
+    # Default virtaul memory base address of architectures.
+    #
+    # This address may different by using different linker.
```
Comment (hanhanW) [inline lib/pwnlib/asm.rb]: may be different

File: test/asm_test.rb [hunk 4]
```diff
@@ -103,4 +114,66 @@ def test_require
...
   end
+
+  def make_elf_file(*args)
+    elf = Asm.make_elf(*args)
+    stream = StringIO.new(elf)
+    [elf, ::ELFTools::ELFFile.new(stream)]
+  end
+
+  def test_make_elf
+    # create elf and use ELFTools to test it
```
Comment (hanhanW) [inline test/asm_test.rb]: create ELF

File: test/asm_test.rb [hunk 5]
```diff
@@ -103,4 +114,66 @@ def test_require
...
+      err = assert_raises(::Pwnlib::Errors::UnsupportedArchError) do
+        Asm.make_elf('')
+      end
+      assert_equal('Unknown machine type of architecture "vax".', err.message)
+    end
+  end
+
+  # this test can be removed after we implement method +run_shellcode+
+  def test_make_elf_and_run
+    # run the elf we created to make sure it works.
```
Comment (hanhanW) [inline test/asm_test.rb]: run the ELF

**Comment by hanhanW** [comment]
:lgtm:

**Pull request merged**